pennylane/tape/tapes/qubit_param_shift.py

```diff
@@ -271,3 +272,85 @@ def processing_fn(results):
             return np.where(self.var_mask, pdA2 - 2 * self._evA_result * pdA, pdA)
 
         return tapes, processing_fn
+
+    def parameter_shift_hessian(self, i, j, params, **options):
+        """Generate the tapes and postprocessing methods required to compute the
+        second derivative with respect to tape parameter :math:`i` and :math:`j`
+        using the second-order parameter-shift method.
+
+        Args:
+            i (int): trainable parameter index to differentiate with respect to
+            j (int): trainable parameter index to differentiate with respect to
+            params (list[Any]): the quantum tape operation parameters
+
+        Keyword Args:
+            s1=pi/2 (float): the size of the shift for index i in the parameter-shift Hessian computations
+            s2=pi/2 (float): the size of the shift for index j in the parameter-shift Hessian computations
+
+        Returns:
+            tuple[list[QuantumTape], function]: A tuple containing the list of generated tapes,
+            in addition to a post-processing function to be applied to the evaluated
+            tapes.
+        """
+        idxs = (i, j)
+        idx_shifts = {i: options.get("s1", np.pi / 2), j: options.get("s2", np.pi / 2)}
+
+        param_shifts = []
+
+        if i == j and idx_shifts[i] == idx_shifts[j]:
+            if idx_shifts[i] == np.pi / 2:
+                # when i = j and s1 = s2 = pi/2, the Hessian parameter shift rule
+                # can be simplified to two device executions
+                param_shifts = [(0.5, 1, np.pi), (-0.5, 1, 0)]
+            elif idx_shifts[i] == np.pi / 4:
+                # when i = j and s1 = s2 = pi/4, the Hessian parameter shift rule
+                # can be simplified to three device executions
+                param_shifts = [(0.5, 1, np.pi / 2), (-1, 1, 0), (0.5, 1, -np.pi / 2)]
+
+        coeffs = []
+        tapes = []
+        shift = np.eye(len(params))
+
+        if param_shifts:
+            # optimizations can be made to reduce amount of tape executions
+            for c, a, s in param_shifts:
+                shifted_tape = self.copy(copy_operations=True, tape_cls=QuantumTape)
+                shifted_tape.set_parameters(a * params + s * shift[i])
+
+                coeffs.append(c)
+                tapes.append(shifted_tape)
+        else:
+            # no optimizations can be made, generate all 4 tapes
+            for idx in idxs:
+                t_idx = list(self.trainable_params)[idx]
+                op = self._par_info[t_idx]["op"]
+                p_idx = self._par_info[t_idx]["p_idx"]
+                s = idx_shifts[idx]
+
+                param_shift = op.get_parameter_shift(p_idx, shift=s)
+                param_shifts.append(param_shift)

```

Comment:
I _think_ this for loop could be moved to line 309 as an `else` branch:

```python
if i == j and idx_shifts[i] == idx_shifts[j]:
    # optimize the number of parameter shifts
    if idx_shifts[i] == np.pi / 2:
        ...
    elif idx_shifts[i] == np.pi / 4:
        ...
else:
    # general 4-term second-derivative shift
    for idx in idxs:
        t_idx = list(self.trainable_params)[idx]
        op = self._par_info[t_idx]["op"]
        p_idx = self._par_info[t_idx]["p_idx"]
        s = idx_shifts[idx]

        param_shift = op.get_parameter_shift(p_idx, shift=s)
        param_shifts.append(param_shift)
```

Reply:
This has the (potential) advantage that the loop below generating the shifts can be used for any form of the `param_shifts` list. Although I might be missing something.

Reply:
I thought of this, but I think this fails when `i == j and idx_shifts[i] == idx_shifts[j]` is true, but `idx_shifts[i]` is not `np.pi / 2` or `np.pi / 4`. That would require another `else` branch inside the first `if` which would do the same thing as the outer else branch.

Reply:
Got it!